RecoTauTag/RecoTau/plugins/DeepTauIdSonicProducer.cc

```diff
@@ -0,0 +1,570 @@
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "HeterogeneousCore/SonicTriton/interface/TritonEDProducer.h"
+#include "RecoTauTag/RecoTau/interface/DeepTauBase.h"
+#include "RecoTauTag/RecoTau/interface/DeepTauHelper.h"
+
+namespace deep_tau {
+  constexpr int NumberOfOutputs = 4;
+}
+
+using namespace deep_tau_2017;
+using namespace deeptau_helper;
+
+class DeepTauIdSonicProducer : public TritonEDProducer<> {
+public:
+  explicit DeepTauIdSonicProducer(edm::ParameterSet const& cfg)
+      : TritonEDProducer<>(cfg),
+        tausToken_(consumes<TauCollection>(cfg.getParameter<edm::InputTag>("taus"))),
```

Comment:
I think there is still a considerable fraction of code duplication, which potentially can be avoided. E.g. parameters.
I propose to extract this common part into DeepTauBase (changing the base class for it to be a template):

```c++
template<typename BaseProducer>
class DeepTauBase : public BaseProducer {};

class DeepTauId : public DeepTauBase<edm::stream::EDProducer<edm::GlobalCache<DeepTauCache>>> {};
class DeepTauIdSonicProducer : public DeepTauBase<TritonEDProducer<>> {};
```